File: saral_hr/saral_hr/report/loan_ledger/loan_ledger.js
// Copyright (c) 2026, sj and contributors
// For license information, please see license.txt

frappe.query_reports["Loan Ledger"] = {
	filters: [
		{
			fieldname: "company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
		},
		{
			fieldname: "loan",
			label: __("Loan"),
			fieldtype: "Link",
			options: "Employee Loan",
			reqd: 1,
			get_query() {
				const company = frappe.query_report.get_filter_value("company");
				const filters = { docstatus: 1 };
				if (company) filters.company = company;
				return { filters };
			},
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
		},
	],

	onload(report) {
		report.page.set_primary_action(__("Print"), () => {
			const f = report.get_values();
			if (!f.loan) {
				frappe.msgprint({
					title: __("Missing Filters"),
					message: __("Please select Loan before printing."),
					indicator: "orange",
				});
				return;
			}
			frappe.dom.freeze(__("Generating PDF…"));
			frappe.call({
				method: "saral_hr.saral_hr.report.loan_ledger.loan_ledger.print_report",
				args: { filters: JSON.stringify(f) },
				callback(r) {
					frappe.dom.unfreeze();
					if (r.message) {
						const a = Object.assign(document.createElement("a"), {
							href: frappe.urllib.get_full_url(r.message),
							target: "_blank",
							rel: "noopener noreferrer",
						});
						document.body.appendChild(a);
						a.click();
						document.body.removeChild(a);
					}
				},
				error() {
					frappe.dom.unfreeze();
					frappe.msgprint({
						title: __("Error"),
						message: __("Failed to generate PDF."),
						indicator: "red",
					});
				},
			});
		}, "printer");
	},

	formatter(value, row, column, data, default_formatter) {
		if (!data) return default_formatter(value, row, column, data);
		const def = (v) => default_formatter(v, row, column, data);
		if (data.bold) {
			if (column.fieldname === "posting_date") return "";
			return `<strong>${def(value)}</strong>`;
		}
		return def(value);
	},
};

File: saral_hr/saral_hr/report/loan_ledger/loan_ledger.py
# Copyright (c) 2026, sj and contributors
# For license information, please see license.txt

import frappe
from frappe import _

from saral_hr.saral_hr.report.loan_ledger_common import (
	build_ledger_pdf_html,
	build_ledger_rows,
	ledger_columns,
	parse_filters,
	save_ledger_pdf,
)


def execute(filters=None):
	filters = filters or {}
	columns = ledger_columns()
	if not filters.get("loan"):
		return columns, []

	rows, _totals = build_ledger_rows(
		loan=filters.get("loan"),
		company=filters.get("company"),
		from_date=filters.get("from_date"),
		to_date=filters.get("to_date"),
	)
	return columns, rows


@frappe.whitelist()
def print_report(filters):
	filters = parse_filters(filters)
	if not filters.get("loan"):
		frappe.throw(_("Loan is required to print the ledger."))

	rows, _totals = build_ledger_rows(
		loan=filters.get("loan"),
		company=filters.get("company"),
		from_date=filters.get("from_date"),
		to_date=filters.get("to_date"),
	)
	loan = frappe.db.get_value(
		"Employee Loan",
		filters.get("loan"),
		["employee", "full_name", "company", "amount"],
		as_dict=True,
	) or {}
	company = filters.get("company") or loan.get("company") or ""
	meta = [
		"Loan: {0}".format(filters.get("loan")),
		"Employee: {0} ({1})".format(loan.get("employee") or "", loan.get("full_name") or ""),
	]
	if filters.get("from_date") or filters.get("to_date"):
		meta.append(
			"Period: {0} to {1}".format(
				filters.get("from_date") or "…", filters.get("to_date") or "…"
			)
		)
	html = build_ledger_pdf_html("Loan Ledger", company, meta, rows)
	return save_ledger_pdf(html, "Loan_Ledger")
